fs-storage/src/file_storage.rs

```diff
@@ -0,0 +1,246 @@
+use std::fmt::Debug;
+use std::fs::{self, File};
+use std::io::{BufRead, BufReader, BufWriter, Write};
+use std::str::FromStr;
+use std::time::SystemTime;
+use std::{
+    collections::HashMap,
+    fmt::Display,
+    path::{Path, PathBuf},
+};
+
+use data_error::{ArklibError, Result};
+
+const LOG_PREFIX: &str = "[file-storage]";
+const STORAGE_VERSION: i32 = 2;
+const STORAGE_VERSION_PREFIX: &str = "version ";
+const KEY_VALUE_SEPARATOR: char = ':';
+
+pub struct FileStorage {
+    log_prefix: String,
+    label: String,
+    path: PathBuf,
+    timestamp: SystemTime,
+}
+
+impl FileStorage {
+    /// Create a new file storage with a diagnostic label and file path
+    pub fn new(label: String, path: &Path) -> Self {
+        Self {
+            label: label.clone(),
+            log_prefix: format!("{} {}", LOG_PREFIX, label).to_string(),
+            path: PathBuf::from(path),
+            timestamp: SystemTime::now(),
+        }
+    }
+
+    /// Read data from disk
+    ///
+    /// Data is read as a key value pairs separated by a symbol and stored
+    /// in a [HashMap] with a generic key K and V value. A handler
+    /// is called on the data after reading it.
+    pub fn read_file<K, V>(
+        &self,
+        mut handle: impl FnMut(HashMap<K, V>),
+    ) -> Result<()>
```

Comment:
I don't remember exact reason why it was implemented this way in Kotlin, but probably this time we can start with simpler and cleaner API and merge `read_file` into `read_file_from_disk`. So, there should be only function returning a mapping (potentially wrapped into Storage value).

Reply:
>  It'll be helpful to know why it was implemented callback style in Kotlin. So that later the design doesn't cause issue.

Agree, but it seems to be too difficult at this moment.. This time, we'll create core components basing on common sense and then adjust existing apps to them.